### PR TITLE
Push primary term to replication tracker

### PR DESCRIPTION
**Pull request by jasontedor**
This commit pushes the primary term into the replication tracker. This is a precursor to using the primary term to resolving ordering problems for retention leases. Namely, it can be that out-of-order retention lease sync requests arrive on a replica. To resolve this, we need a tuple of (primary term, version). For this to be, the primary term needs to be accessible in the replication tracker. As the primary term is part of the replication group anyway, this change conceptually makes sense.

Relates #37951
Relates #38036

### CODE REVIEW

**Review by dnhatn — COMMENTED** [review]
> As the primary term is part of the replication group anyway, this change conceptually makes sense.

This makes sense to me too. I thought we are going to provide a PrimaryTermSupplier to ReplicationTracker and I find that having `pendingPrimaryTerm` and `operationPrimaryTerm` at the same place is a bit easier to understand than two places. 

This change looks good to me, but I prefer to wait for @ywelsch on this.

File: server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java [hunk 1]
```diff
@@ -92,6 +93,31 @@
...
+     * {@link #setOperationPrimaryTerm(long)}.
+     */
+    private volatile long operationPrimaryTerm;
+
+    /**
+     * Returns the current operation primary term.
+     *
+     * @return the primary term
+     */
+    public long getOperationPrimaryTerm() {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java]: please move getter and setter further down. I understand this is your preferred coding style. However, it conflicts with the coding style we have adapted for all these classes and don't think this is the right time and place to switch things up.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -336,8 +336,9 @@ public boolean shouldCache(Query query) {
         }
         indexShardOperationPermits = new IndexShardOperationPermits(shardId, threadPool);
         searcherWrapper = indexSearcherWrapper;
-        pendingPrimaryTerm = indexSettings.getIndexMetaData().primaryTerm(shardId.id());
-        operationPrimaryTerm = pendingPrimaryTerm;
+        final long primaryTerm = indexSettings.getIndexMetaData().primaryTerm(shardId.id());
+        this.pendingPrimaryTerm = primaryTerm;
+        replicationTracker.setOperationPrimaryTerm(primaryTerm);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: can we set this in the constructor of ReplicationTracker? Then we would always have a proper operation term in ReplicationTracker.

**Review by ywelsch — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java (hunk 1 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java]: I pushed 72e2e7b15e7cc3d981f956c8b015e92ba7ccf19d.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 2 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I pushed 72e2e7b15e7cc3d981f956c8b015e92ba7ccf19d.

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/1

**Pull request merged**